File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenSubscriber.java
package no.bankaxept.epayment.client.base.accesstoken;

import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Flow;
import java.util.concurrent.TimeUnit;
import no.bankaxept.epayment.client.base.AccessFailed;
import no.bankaxept.epayment.client.base.accesstoken.publisher.AccessTokenPublisher;

public class AccessTokenSubscriber implements Flow.Subscriber<String> {

  private final CompletableFuture<String> token = new CompletableFuture<>();

  public AccessTokenSubscriber(AccessTokenPublisher tokenProcessor) {
    tokenProcessor.subscribe(this);
  }

  @Override
  public void onSubscribe(Flow.Subscription subscription) {
  }

  @Override
  public void onNext(String item) {
    token.complete(item);
  }

  @Override
  public void onError(Throwable throwable) {
    token.completeExceptionally(throwable);
  }

  @Override
  public void onComplete() {
  }

  public String get(Duration timeout) {
    try {
      return token.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
    } catch (ExecutionException e) {
      throw new AccessFailed(e.getCause());
    } catch (Exception e) {
      throw new AccessFailed(e);
    }
  }
}

File: csharp/BaseClient/src/AccessToken/AccessTokenSubscriber.cs
using System;
using System.Threading.Tasks;

namespace BankAxept.Epayment.Client.Base.AccessToken {
    public class AccessTokenSubscriber : IObserver<string> {

        private readonly TaskCompletionSource<string> token = new TaskCompletionSource<string>();

        public AccessTokenSubscriber(IObservable<string> tokenPublisher) {
            tokenPublisher.Subscribe(this);
        }

        public void OnCompleted() {}

        public void OnError(Exception error) {
            token.TrySetException(error);
        }

        public void OnNext(string value) {
            token.TrySetResult(value);
        }

        public string Get(TimeSpan timeout) {
            try {
                if (token.Task.Wait(timeout)) {
                    return token.Task.Result;
                } else {
                    throw new AccessFailed(new TimeoutException("Timeout waiting for token"));
                }
            }
            catch (AggregateException e) {
                throw new AccessFailed(e.InnerException);
            }
            catch (Exception e){
                throw new AccessFailed(e);
            }
        }
    }
}